---
title: "Explore a gene expression data set"
teaching: 45
exercises: 0
questions:
- "How do I get an overview of an RNA-seq dataset?"
- "How do I calculate differentially expressed genes?"
- "How do I find enriched biological processes?"
objectives:
- "Get introduced to using a specific R-package."
- "Learn how to perform basic gene expression analysis."
keypoints:
- "Use vignettes to get introduced to a new package."
- "Gene expression analysis edgeR can be quite straight-forward."
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("04-explore-gene-expression-")
```


von Wulffen et al has deposited a [RNA-seq expression dataset](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE71562) from studying the effects on *E. coli* transitioning from anaerobic conditions to aerobic conditions. Three biological replicate cultures where grown in anaerobic conditions, sampled, then subjected to aeration at 1 l/min and new samples were taken after 0.5, 1, 2, 5 and 10   min. Total RNA was extracted from the samples, ribo-depleted and sequenced on Illumina HISeq. Reads were aligned to K12 reference genome and counted for each gene. 

That data has been downloaded here and we will here use it to provide an example of how to perform a introductory analysis using the edgeR package. We will

- read the data to R
- perform a PCA (principal component analysis) to get an
  overview of how dissimilar the samples are
- find genes that are up/down regulated upon aeration
- figure out which biological processes are affected mostly

## Installing bioconductor packages <!-- 3 -->

We will make use of the Bioconductor `edgeR` package as well as the `org.EcK12.eg.db` package so we start by downloading and installing those.

> ## Why not install.packages?
> 
> Bioconductor was launched as collection of bioinformatics related packages that used the same core classes and worked well together. To enable this, they wanted to have their own repository, subjected to special checks and review process and therefore decided to not use the regular CRAN repository. You can still use `install.packages` to get Bioconductor packages but then have to specify the repository manually. :/
{: .callout}

```{r, eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite(c("edgeR", "org.EcK12.eg.db"))
```

Then load the required packages.

```{r, results="hide", message=FALSE, warning=FALSE}
library(edgeR)
library(ggplot2)
library(org.EcK12.eg.db)
```

`edgeR` comes with very good user manual. You can access it by

```{r, eval=FALSE}
edgeRUsersGuide()
```

> ## Vignettes often provide great introduction to packages
> `edgeR` provides a special function to open the vignette, other packages use the `vignette(topic, package)` function. See which vignettes are available for e.g. ggplot2!
{: .callout}


## Read the data <!-- 3 -->

The read-counts data is simply a table and we already know how to read those.

```{r}
wulffenTable <- read.table("data/GSE71562.csv", header=TRUE, row.names=1, sep=",")
head(wulffenTable)
```

Genes in rows, samples in columns. 

We also need to know which sample is which is which and there is a different file that contains that information.

```{r}
samples <- read.table("data/pheno.csv", header=TRUE, row.names=1, sep=",")
samples
```

We then create a `DGEList` which is a class used by `edgeR` and calculate normalization factor for each library (to make sure we don't overestimate expression of genes that come from samples that were sequenced deeper).

```{r}
wulffen <- DGEList(counts=wulffenTable, genes=rownames(wulffenTable), samples=samples)
wulffen <- calcNormFactors(wulffen)
```

## Exploring the data <!-- 10 -->

An often very useful way to explore large datasets is to perform a PCA and plot the samples in 2D that maximally capture the variation in the dataset. This must be done on a statistic for each gene that is independent on the length of the gene so for this purpose we calculate get the 'counts per million' matrix.

```{r}
wulffenCpm <- cpm(wulffen)
```

Then we perform PCA using the `pcaMethods` package.

```{r, message=FALSE, warning=FALSE}
library(pcaMethods)
pcaFit <- pca(log2(t(wulffenCpm) + 0.25))
pcaFit
```

> ## What did the `t` do? Why `+ 0.25`?
>
> `prcomp` requires the `variables` in this case the genes, to come in
> the rows so we used `t` to transpose the data matrix. Since we know
> gene expression values tend to follow log-normal distributions, we
> use `log2` to transform the data. Why did we add the magic value
> `0.25`? Try removing it and see what you get.
> {: .r}
{: .challenge}

To get a nice data frame that we can use for plotting we simply use `merge` with the samples data fram.

```{r}
pcaDf <- merge(scores(pcaFit), samples, by=0)
```

Then we can plot the data using `ggplot2`

```{r}
ggplot(pcaDf, aes(PC1, PC2, label=time, color=replicate)) +
    geom_text()
```    

or perhaps

```{r}
ggplot(pcaDf, aes(PC1, PC2, label=time, color=replicate)) +
    geom_text(nudge_x=1) +
    geom_path() +
    geom_point()
```    

The time-series can easily be recognized which is a good sign that experiment was successful.

> ## More PCA dimensions?
>
> It is often useful to look at more than 2 dimensions in pca. Read the documentation for `pca` to compute more principal components, and visualize e.g. PC2 and PC3 or PC1 and PC5. Any new conclusions?
> {: .r}
{: .challenge}

Scatter plots are also a great way to FIXME

## Differentially expressed genes <!-- 10 -->

From our PCA we could, as expected, see that the last timepoint is the most dissimilar from the the anaerobic condition. Let's make a comparison between the anaerobic and 10 min anearobic samples and see which genes are differentially expressed between those.

With `edgeR` we will fit a simple generalized linear model to get estimates for differential expression and for that we first need to create a *design matrix* that accurately describes the comparison we are after.

```{r}
wulffenShort <- wulffen[, wulffen$samples$time %in% c("t0", "t10")]
design <- model.matrix(~as.character(time), data=wulffenShort$samples)
colnames(design) <- c("(Intercept)", "t10")
design
```

The matrix we just created indicates which samples should be used to calculate the intercept (all samples) and then the effect of 10 min aeration (the t10 samples). With these objects we can now perform our differential expression analysis. 

```{r}
wulffenShort <- estimateDisp(wulffenShort, design)
fit <- glmFit(wulffenShort, design)
lrt <- glmLRT(fit)
topTags(lrt)
```

What did we just do? The `estimateDisp` function is needed to estimate variance components robustly, `glmFit` fits the model we are after that essentially has one overall mean of expression and another mean for the t10 samples. glmLRT performes a log-likelihood ratio test against the null-hypothesis that t10 has the same average as all the samples together. Then with `topTags` we extract a table with the 10 most differentially expressed genes.

> ## Write the expression estimates to a file
>
> It is often useful to export the data for use in other programs and sharing with colleagues. Use the `write.table` function to export a comma separated file with the output of `topTags` for all genes.
> {: .r}
{: .challenge}

> ## Extra challenge: identifier mappings
>
> Use the `merge` function and the mapping to bnumbers in `data/ecoli.csv` to also add a column with bnumbers as identifiers
> 
> > ## Solution
> > ```{r}
> > df <- merge(topTags(lrt, n=Inf), read.csv("data/ecoli.csv"))
> > write.csv(df, file="data/deg.csv", quote=FALSE, row.names=FALSE)
> ```
> {: .solution}
{: .challenge}

> ## edgeR for RNA-seq, limma for general differential gene expression
>
> If you don't have read counts, but e.g. abundance estimates for proteomics, use the `limma` package instead. Syntax and usage is very similar to edgeR, read the `limmaUsersGuide()`!
> {: .r}
{: .callout}
## Short introduction to annotation packages <!-- 5 -->
A great assset in Bioconductor are the annotation packages which makes differential gene expression analysis quite straight-forward (if you work on the most common model species..). The most relevant packages are the *organism annotation packages* and they are all named `org.`, then a species code, the main identifier, and `.db`. For E. coli K-12, entrez gene identifiers, it is thus `org.EcK12.db`. To get an overview of what is in an annotation package call the package name minus the `.db` as a function.

```{r}
library(org.EcK12.eg.db)
org.EcK12.eg()
```

One can convert the objects to lists directly

```{r}
head(as.list(org.EcK12.egSYMBOL))
```

or use the `mapIds` function

```{r}
mapIds(org.EcK12.eg.db, 'yaaJ', 'ENTREZID', keytype='SYMBOL')
```

> ## Simple linkout to a paper
>
> Which paper mentions the yaaJ gene?
> 
> > ## Solution
> > ```{r, eval=FALSE}
> > pmid <- mapIds(org.EcK12.eg.db, 'yaaJ', 'PMID', keytype='SYMBOL')[1]
> > browseURL(paste0('https://www.ncbi.nlm.nih.gov/pubmed/?term=', pmid))
> > ```
> {: .solution}
{: .challenge}

## Over-representation analysis of biological processes <!-- 10 -->
We want to examine if the most differentially expressed genes have any particular biological processes in common. We will do this using the function `gage` from the `gage` package to perform Gene Set Enrichment Analysis (GSEA). The input to `gage` must be Entrez identifiers so we first need to map our gene symbols to Entrez. Bioconductor conveniently provides this mapping so all we need to do is to load the right annotation package and map our identifiers. We also define an object `universe` which holds all the genes which were present in our dataset, and which could be mapped to Entrez identifiers - that is simply all our mapped genes except the missing values (na = not available).

First we need to load some packages that we are going to use.

```{r, message=FALSE, warning=FALSE}
library(org.EcK12.eg.db)
library(GO.db)
library(gage)
```

```{r}
symbol2entrez <- mapIds(org.EcK12.eg.db, rownames(lrt), "ENTREZID", keytype="SYMBOL")
fc <- lrt$table$logFC[!is.na(symbol2entrez)]
names(fc) <- na.omit(symbol2entrez)
```

Next we need to create a mapping between gene identifiers and GO terms.

```{r}
allGo <- as.list(GOTERM)
bp <- Filter(function(go) go@Ontology == 'BP', allGo)
goMapping <- mget(names(bp), org.EcK12.egGO2ALLEGS, ifnotfound=NA)
```

Then we do the gene set enrichment analysis

```{r}
goFc <- gage(fc, goMapping)
goGreater <- as.data.frame(goFc$greater)
goGreater <- goGreater[goGreater$q.val < 0.01 & !is.na(goGreater$q.val),]
goTermNames <- lapply(mget(rownames(goGreater), GOTERM), function(go) go@Term)
goGreater$Term <- goTermNames[rownames(goGreater)]
goGreater
```

> ## Which genes decreased in transition to aerobic conditions?
>
> Adjust the code to instead look a genes that decreased in expression!
{: .challenge}
